src/main/java/com/google/firebase/FirebaseOptions.java

```diff
@@ -69,11 +70,19 @@
   private final ThreadManager threadManager;
   private final FirestoreOptions firestoreOptions;
 
```

Comment:
I've tried not to break the public API for `FirebaseOptions.Builder`: so it supports both `setDeferredCredentials` (which is package private, and can't be used publicly) and the old `setCredentials` method. This constructor accepts `defaultCredentials` if not null, else falls back to creating a custom `DefaultApplicationCredentials` wrapping the regular `credentials`.

Suggestions welcome for doing this in a better way.

Reply:
As eluded to above, you essentially re-implemented the Supplier  pattern (available via com.google.common.base.Supplier in Java 7). I think this could just be:

```
Supplier<Credentials> credentials;

public void setCredentials(Credentials credentials) { this.credentials = Suppliers.ofInstance(credentials); }

void setCredentials(Supplier<Credentials> credentials) { this.credentials = credentials; }

```